File: bms/go/go-gt/Matrix_TransitiveClosure/Matrix_TransitiveClosure.go
package test

var ()

const ()

func (G *Matrix) TransitiveClosure(N *Matrix) {
	var i, j, k int64
	for i = 0; i < G.N; i++ {
		for j = 0; j < G.N; j++ {
			for k = 0; k < G.N; k++ {
				if G.Get(i, k) > 0 && G.Get(k, j) > 0 {
					if G.Get(i, j) == 0 || G.Get(i, k)+G.Get(k, j) < G.Get(i, j) {
						G.Set(i, j, G.Get(i, k)+G.Get(k, j))
						if N != nil {
							N.Set(i, j, k+1)
						}
					}
				}
			}
		}
	}
}

func (m Matrix) Get(i int64, j int64) int64 {
	return m.A[i*m.N+j]
}

func (m Matrix) Set(i int64, j int64, v int64) {
	m.A[i*m.N+j] = v
}

type Matrix struct {
	N int64
	A []int64
}

File: translations_only/fail/go/go-gt/Matrix_TransitiveClosure/gemini/CAPR/Matrix_TransitiveClosure.rs

pub fn transitive_closure(g: &mut Matrix, n: &mut Matrix) {
    for i in 0..g.n {
        for j in 0..g.n {
            for k in 0..g.n {
                if g.get(i, k) > 0 && g.get(k, j) > 0 {
                    if g.get(i, j) == 0 || g.get(i, k) + g.get(k, j) < g.get(i, j) {
                        g.set(i, j, g.get(i, k) + g.get(k, j));
                        if !n.empty() {
                            n.set(i, j, k + 1);
                        }
                    }
                }
            }
        }
    }
}

impl Matrix {
    pub fn get(&self, i: i64, j: i64) -> i64 {
        self.a[i as usize * self.n as usize + j as usize]
    }

    pub fn set(&mut self, i: i64, j: i64, v: i64) {
        self.a[i as usize * self.n as usize + j as usize] = v;
    }

    pub fn empty(&self) -> bool {
        self.n == 0
    }
}

pub struct Matrix {
    pub n: i64,
    pub a: Vec<i64>,
}
